content/en/blog/2023/egress-sni/index.md

```diff
@@ -0,0 +1,441 @@
+publishdate: 2023-12-10
+attribution: "Gergő Huszty (IBM)"
+keywords: [traffic-management,gateway,mesh,mtls,egress,remote]
+---
+
+If you are using Istio to handle application-originated traffic to destinations outside of the mesh, you're probably familiar with the concept of egress gateways.
+Egress gateways can be used to monitor and forward traffic from mesh-internal applications to locations outside of the mesh.
+This is a useful feature if your system is operating in a restricted
+environment and you want to control what can be reached on the public internet from your mesh.
+
+## Background
+
+The use-case of configuring an egress gateway to handle arbitrary wildcard domains had been included in the [official Istio docs](/docs/tasks/traffic-management/egress/wildcard-egress-hosts/#wildcard-configuration-for-arbitrary-domains) up until version 1.13, but was subsequently removed because the documented solution was not officially supported or recommended and was subject to breakage in future versions of Istio.
+Nevertheless, the old solution was still usable with Istio versions before 1.20. Istio 1.20, however, dropped some Envoy functionality that was required for the approach to work.
+
+This post attempts to describe how we resolved the issue and filled the gap with a similar approach using Istio version-independent components and Envoy features, but without the need for a separate Nginx SNI proxy.
+Our approach allows users of the old solution to seamlessly migrate configurations before their systems face the breaking changes in Istio 1.20.
+
+## Problem to solve
+
+The currently documented egress traffic routing use-cases are all building on the assumption that the target of the traffic
+(the hostname) is statically determined by a `VirtualService`, that tells Envoy in the egress gateway pod where to TCP proxy
+the matching outbound connections. You can use multiple and even wildcard DNS names to match the routing criteria, but you
+are not able to route the traffic exactly where the application wanted to. For example you can match traffic that targets
+`*.wikipedia.org`, but you can direct everything into a single final target, i.e. `en.wikipedia.org`. If there is a
+service on `anyservice.wikipedia.org` that is not hosted by the same server(s) as `en.wikipedia.org`, your traffic to that
+arbitrary host will fail. In fact the connection would go to `en.wikipedia.org` servers, while the target hostname in the
+TLS handshake and the HTTP payload would contain `anyservice.wikipedia.org`, which may not be served by the same servers.
+The solution on high level is to inspect the original server name (SNI extension) in the application TLS handshake (that is sent
+in plain-text, so no TLS termination or other man-in-the-middle operation is needed) in every new connection and use it as
+a target dynamically to TCP proxy the traffic when leaves the gateway.
```

Comment:
```suggestion
The currently documented egress gateway use-cases rely on the fact that the target of the traffic
(the hostname) is statically configured in a `VirtualService`, telling Envoy in the egress gateway pod where to TCP proxy
the matching outbound connections. You can use multiple, and even wildcard, DNS names to match the routing criteria, but you
are not able to route the traffic to the exact location specified in the application request. For example you can match traffic for targets
`*.wikipedia.org`, but you then need to forward the traffic to a single final target, e.g., `en.wikipedia.org`. If there is another
service, e.g., `anyservice.wikipedia.org` that is not hosted by the same server(s) as `en.wikipedia.org`, traffic to that host will fail because, even though the target hostname in the
TLS handshake of the HTTP payload contains `anyservice.wikipedia.org`, the `en.wikipedia.org` servers will not be be able to serve the request.

The solution to this problem at a high level is to inspect the original server name (SNI extension) in the application TLS handshake (which is sent
in plain-text, so no TLS termination or other man-in-the-middle operation is needed) in every new gateway connection and use it as
the target to dynamically TCP proxy the traffic leaving the gateway.
```